x/denommetadata/ibc_middleware.go

```diff
@@ -0,0 +1,134 @@
+package denommetadata
+
+import (
+	"errors"
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	errortypes "github.com/cosmos/cosmos-sdk/types/errors"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+
+	"github.com/dymensionxyz/dymension-rdk/x/denommetadata/types"
+)
+
+var _ porttypes.IBCModule = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the transfer middleware
+type IBCMiddleware struct {
+	porttypes.IBCModule
+	keeper DenomMetadataKeeper
+}
+
+type DenomMetadataKeeper interface {
+	HasDenomMetaData(ctx sdk.Context, denom string) bool
+	CreateDenomMetadata(ctx sdk.Context, metadatas ...types.DenomMetadata) error
+}
+
+// NewIBCMiddleware creates a new IBCMiddleware given the keeper and underlying application
+func NewIBCMiddleware(keeper DenomMetadataKeeper, app porttypes.IBCModule) IBCMiddleware {
+	return IBCMiddleware{
+		IBCModule: app,
+		keeper:    keeper,
+	}
+}
+
+// OnRecvPacket registers the denom metadata if it does not exist
+func (im IBCMiddleware) OnRecvPacket(
```

Comment:
how do you know the packet is from the hub?

Reply:
Sorry if it's obvious
I didn't see it
In any case, if it's secure, needs a comment explaining